File: backend/src/bond_monitor/domain/portfolio/cashflow.py
"""Cashflow-события плана и слияние дублирующихся строк timeline."""

from __future__ import annotations

from collections.abc import Callable, Sequence
from dataclasses import dataclass
from datetime import date

from bond_monitor.domain.portfolio.models import ReinvestmentSlot, ReinvestmentTriggerReason


@dataclass
class CashflowEvent:
    """Атомарное событие денежного потока в плане портфеля.

    Знак ``amount_rub``:
        * положительный → приток денег в кэш-баланс (купон, погашение, оферта);
        * отрицательный → отток (покупка бумаги).

    ``is_projected = True`` означает, что событие лежит в будущем и
    основано на текущих рыночных параметрах; история (если в портфеле есть
    позиции, купленные в прошлом) идёт с ``is_projected = False``.
    """

    date: date
    kind: str
    amount_rub: float
    description: str
    related_isin: str | None = None
    is_projected: bool = True
    position_id: int | None = None
    lots: int | None = None
    bonds_count: int | None = None


def event_sort_key(event: CashflowEvent) -> tuple[date, int]:
    """Сортировка событий: внутри одной даты сначала покупки, потом купоны/погашения."""
    order = {"purchase": 0, "coupon": 1, "maturity": 2, "put_offer": 2}
    return (event.date, order.get(event.kind, 3))


def _format_bonds_count_suffix(bonds_count: int | None) -> str:
    if bonds_count is None or bonds_count <= 0:
        return ""
    return f" ({bonds_count} шт.)"


def cashflow_event_description(
    kind: str,
    name: str,
    *,
    bonds_count: int | None,
    lots: int | None = None,
    price_suffix: str = "",
) -> str:
    suffix = _format_bonds_count_suffix(bonds_count)
    if kind == "purchase":
        return f"Покупка {lots} лот(а) — {name}{suffix}"
    if kind == "coupon":
        return f"Купон по {name}{suffix}"
    if kind == "put_offer":
        return f"Пут-оферта по {name}{price_suffix}{suffix}"
    return f"Погашение {name}{suffix}"


def _bond_name_from_cashflow_description(description: str) -> str:
    text = description
    if text.endswith(" шт.)"):
        text = text.rsplit(" (", 1)[0]
    if " — " in text:
        return text.split(" — ", 1)[1]
    for prefix in ("Купон по ", "Погашение ", "Пут-оферта по "):
        if text.startswith(prefix):
            return text[len(prefix) :]
    return text


_MERGEABLE_EVENT_KINDS = frozenset({"coupon", "maturity", "put_offer", "purchase"})


def _refresh_merged_cashflow_description(event: CashflowEvent) -> None:
    """Пересобрать описание после слияния событий с суммированным количеством."""
    if not event.related_isin:
        return
    name = _bond_name_from_cashflow_description(event.description)
    if event.kind == "purchase":
        event.description = cashflow_event_description(
            "purchase",
            name,
            bonds_count=event.bonds_count,
            lots=event.lots,
        )
        return
    price_suffix = ""
    if event.kind == "put_offer" and " (" in event.description:
        tail = event.description.split(" (", 1)[1]
        if "% номинала)" in tail:
            price_suffix = f" ({tail.split(')')[0]})"
    event.description = cashflow_event_description(
        event.kind,
        name,
        bonds_count=event.bonds_count,
        price_suffix=price_suffix,
    )


def merge_cashflow_events(events: list[CashflowEvent]) -> list[CashflowEvent]:
    """Объединить события одной бумаги в один день в одну строку cashflow.

    Несколько позиций (initial + phantom-ы реинвестиций) с одним ISIN
    эмитят отдельные купоны/погашения/покупки на одну дату — для UI
    суммируем их в одно событие.
    """
    sorted_input = sorted(events, key=event_sort_key)
    merged: dict[tuple[date, str, str], CashflowEvent] = {}
    merge_order: list[tuple[date, str, str]] = []
    passthrough: list[CashflowEvent] = []

    for event in sorted_input:
        if event.kind not in _MERGEABLE_EVENT_KINDS or not event.related_isin:
            passthrough.append(event)
            continue
        key = (event.date, event.kind, event.related_isin)
        existing = merged.get(key)
        if existing is None:
            merged[key] = CashflowEvent(
                date=event.date,
                kind=event.kind,
                amount_rub=event.amount_rub,
                description=event.description,
                related_isin=event.related_isin,
                is_projected=event.is_projected,
                lots=event.lots,
                bonds_count=event.bonds_count,
            )
            merge_order.append(key)
            continue
        existing.amount_rub += event.amount_rub
        existing.is_projected = existing.is_projected or event.is_projected
        if event.lots is not None:
            existing.lots = (existing.lots or 0) + event.lots
        if event.bonds_count is not None:
            existing.bonds_count = (existing.bonds_count or 0) + event.bonds_count
        _refresh_merged_cashflow_description(existing)

    result = [merged[key] for key in merge_order] + passthrough
    result.sort(key=event_sort_key)
    return result


def _slot_sort_key(slot: ReinvestmentSlot) -> tuple[date, int, str]:
    """Сортировка слотов реинвестиции по дате наступления события."""
    reason_order = {
        ReinvestmentTriggerReason.MATURITY: 0,
        ReinvestmentTriggerReason.PUT_OFFER: 1,
        ReinvestmentTriggerReason.COUPON_CASH: 2,
    }
    return (
        slot.trigger_date,
        reason_order.get(slot.trigger_reason, 3),
        slot.source_position_isin or slot.effective_isin or "",
    )


def _slot_merge_key(slot: ReinvestmentSlot) -> tuple[date, str, str] | tuple[date, str] | None:
    if slot.trigger_reason == ReinvestmentTriggerReason.COUPON_CASH:
        return (slot.trigger_date, slot.trigger_reason.value)
    if slot.source_position_isin:
        return (slot.trigger_date, slot.trigger_reason.value, slot.source_position_isin)
    return None


def _slot_coalesce_key(slot: ReinvestmentSlot) -> tuple[date, str] | None:
    effective_isin = slot.effective_isin
    if not effective_isin:
        return None
    return (slot.purchase_date, effective_isin)


_SLOT_REASON_PRIORITY = {
    ReinvestmentTriggerReason.MATURITY: 0,
    ReinvestmentTriggerReason.PUT_OFFER: 1,
    ReinvestmentTriggerReason.COUPON_CASH: 2,
}


def _copy_reinvestment_slot(slot: ReinvestmentSlot) -> ReinvestmentSlot:
    return ReinvestmentSlot(
        trigger_date=slot.trigger_date,
        trigger_reason=slot.trigger_reason,
        expected_cash_rub=slot.expected_cash_rub,
        suggested_isin=slot.suggested_isin,
        suggested_name=slot.suggested_name,
        confirmed_isin=slot.confirmed_isin,
        gap_days=slot.gap_days,
        source_position_isin=slot.source_position_isin,
    )


def _accumulate_reinvestment_slot(existing: ReinvestmentSlot, slot: ReinvestmentSlot) -> None:
    existing.expected_cash_rub += slot.expected_cash_rub
    if existing.confirmed_isin is None and slot.confirmed_isin is not None:
        existing.confirmed_isin = slot.confirmed_isin
    if existing.suggested_isin is None and slot.suggested_isin is not None:
        existing.suggested_isin = slot.suggested_isin
    if existing.suggested_name is None and slot.suggested_name is not None:
        existing.suggested_name = slot.suggested_name

    existing_priority = _SLOT_REASON_PRIORITY.get(existing.trigger_reason, 99)
    slot_priority = _SLOT_REASON_PRIORITY.get(slot.trigger_reason, 99)
    if slot_priority < existing_priority:
        existing.trigger_reason = slot.trigger_reason
        existing.trigger_date = slot.trigger_date
        if slot.source_position_isin is not None:
            existing.source_position_isin = slot.source_position_isin
    else:
        existing.trigger_date = min(existing.trigger_date, slot.trigger_date)
        if existing.source_position_isin is None and slot.source_position_isin is not None:
            existing.source_position_isin = slot.source_position_isin


def _merge_reinvestment_slot_groups(
    slots: Sequence[ReinvestmentSlot],
    *,
    key_fn: Callable[[ReinvestmentSlot], tuple | None],
) -> list[ReinvestmentSlot]:
    sorted_input = sorted(slots, key=_slot_sort_key)
    merged: dict[tuple, ReinvestmentSlot] = {}
    merge_order: list[tuple] = []
    passthrough: list[ReinvestmentSlot] = []

    for slot in sorted_input:
        key = key_fn(slot)
        if key is None:
            passthrough.append(slot)
            continue
        existing = merged.get(key)
        if existing is None:
            merged[key] = _copy_reinvestment_slot(slot)
            merge_order.append(key)
            continue
        _accumulate_reinvestment_slot(existing, slot)

    result = [merged[key] for key in merge_order] + passthrough
    result.sort(key=_slot_sort_key)
    return result


def merge_reinvestment_slots(slots: list[ReinvestmentSlot]) -> list[ReinvestmentSlot]:
    """Объединить дублирующиеся слоты реинвестиции в одну карточку UI.

    Два прохода:

    1. Одинаковые ``trigger_date`` + ``trigger_reason`` + ``source_position_isin``
       (или ``coupon_cash`` на одну дату) — суммируем phantom-позиции одной бумаги.
    2. Одинаковые ``purchase_date`` + ``effective_isin`` — погашение и купонный
       кэш, которые реинвестируются в одну бумагу в один день, показываем одной
       карточкой.
    """
    by_source = _merge_reinvestment_slot_groups(slots, key_fn=_slot_merge_key)
    return _merge_reinvestment_slot_groups(by_source, key_fn=_slot_coalesce_key)

File: backend/internal/domain/portfolio/cashflow.go
package portfolio

import (
	"sort"
	"strings"
	"time"

	"github.com/tonatos/bond-monitor/backend/internal/domain/shared"
)

// CashflowEvent is an atomic cashflow line in the portfolio plan.
type CashflowEvent struct {
	Date         time.Time
	Kind         string
	AmountRub    float64
	Description  string
	RelatedISIN  *string
	IsProjected  bool
	PositionID   *int64
	Lots         *int
	BondsCount   *int
	JournalSeq   int
}

var causalKindOrder = map[string]int{"coupon": 0, "maturity": 1, "put_offer": 1, "purchase": 2}

func EventSortKey(e CashflowEvent) (time.Time, int) {
	order := map[string]int{"purchase": 0, "coupon": 1, "maturity": 2, "put_offer": 2}
	return e.Date, order[e.Kind]
}

func JournalSortKey(e CashflowEvent) (time.Time, int) {
	if e.JournalSeq > 0 {
		return e.Date, e.JournalSeq
	}
	return e.Date, causalKindOrder[e.Kind]
}

func sortedJournal(events []CashflowEvent) []CashflowEvent {
	out := append([]CashflowEvent(nil), events...)
	sort.Slice(out, func(i, j int) bool {
		di, oi := JournalSortKey(out[i])
		dj, oj := JournalSortKey(out[j])
		if di.Equal(dj) {
			return oi < oj
		}
		return di.Before(dj)
	})
	return out
}

// JoinCashflowJournals merges historical and projected journal segments in sort order.
func JoinCashflowJournals(segments ...[]CashflowEvent) []CashflowEvent {
	var combined []CashflowEvent
	for _, seg := range segments {
		combined = append(combined, seg...)
	}
	return sortedJournal(combined)
}

func RunningCashBeforePurchase(events []CashflowEvent, purchaseDate time.Time, initialCash float64) float64 {
	cash := initialCash
	for _, event := range sortedJournal(events) {
		if event.Date.After(purchaseDate) {
			break
		}
		if event.Date.Before(purchaseDate) {
			cash += event.AmountRub
			continue
		}
		if event.Kind == "purchase" {
			break
		}
		cash += event.AmountRub
	}
	return cash
}

func CashOnHandBeforeDate(events []CashflowEvent, asOf time.Time, initialCash float64) float64 {
	cash := initialCash
	for _, event := range sortedJournal(events) {
		if !event.Date.Before(asOf) {
			break
		}
		cash += event.AmountRub
	}
	return cash
}

type CashflowRow struct {
	Date            string
	AmountRub       float64
	Kind            string
	Label           string
	Lots            *int
	BondsCount      *int
	BalanceAfterRub float64
}

func CashflowRowsWithBalance(events []CashflowEvent, initialCash float64) []CashflowRow {
	return CashflowRowsFromDate(events, initialCash, time.Time{})
}

// CashflowProjectedRowsFromToday returns forward-looking plan rows (projected events from as-of date).
func CashflowProjectedRowsFromToday(events []CashflowEvent, initialCash float64, today time.Time) []CashflowRow {
	today = shared.DateOnly(today)
	if today.IsZero() {
		return CashflowRowsWithBalance(events, initialCash)
	}
	running := CashOnHandBeforeDate(events, today, initialCash)
	var rows []CashflowRow
	for _, event := range sortedJournal(events) {
		if event.Date.Before(today) || !event.IsProjected {
			continue
		}
		running += event.AmountRub
		rows = append(rows, CashflowRow{
			Date:            shared.FormatISODate(event.Date),
			AmountRub:       event.AmountRub,
			Kind:            event.Kind,
			Label:           event.Description,
			Lots:            event.Lots,
			BondsCount:      event.BondsCount,
			BalanceAfterRub: round2(running),
		})
	}
	return rows
}

// CashflowRowsFromDate builds display rows from fromDate inclusive; balance starts at cash on that date.
func CashflowRowsFromDate(events []CashflowEvent, initialCash float64, fromDate time.Time) []CashflowRow {
	running := initialCash
	if !fromDate.IsZero() {
		running = CashOnHandBeforeDate(events, fromDate, initialCash)
	}
	var rows []CashflowRow
	for _, event := range sortedJournal(events) {
		if !fromDate.IsZero() && event.Date.Before(fromDate) {
			continue
		}
		running += event.AmountRub
		rows = append(rows, CashflowRow{
			Date:            shared.FormatISODate(event.Date),
			AmountRub:       event.AmountRub,
			Kind:            event.Kind,
			Label:           event.Description,
			Lots:            event.Lots,
			BondsCount:      event.BondsCount,
			BalanceAfterRub: round2(running),
		})
	}
	return rows
}

func round2(v float64) float64 {
	return float64(int(v*100+0.5)) / 100
}

func CashflowEventDescription(kind, name string, bondsCount, lots *int, priceSuffix string) string {
	suffix := formatBondsCountSuffix(bondsCount)
	switch kind {
	case "purchase":
		l := 0
		if lots != nil {
			l = *lots
		}
		return "Покупка " + itoa(l) + " лот(а) — " + name + suffix
	case "coupon":
		return "Купон по " + name + suffix
	case "put_offer":
		return "Пут-оферта по " + name + priceSuffix + suffix
	default:
		return "Погашение " + name + suffix
	}
}

func itoa(v int) string {
	if v == 0 {
		return "0"
	}
	var b [20]byte
	i := len(b)
	n := v
	for n > 0 {
		i--
		b[i] = byte('0' + n%10)
		n /= 10
	}
	return string(b[i:])
}

func formatBondsCountSuffix(bondsCount *int) string {
	if bondsCount == nil || *bondsCount <= 0 {
		return ""
	}
	return " (" + itoa(*bondsCount) + " шт.)"
}

func bondNameFromDescription(description string) string {
	text := description
	if strings.HasSuffix(text, " шт.)") {
		text = strings.Split(text, " (")[0]
	}
	if idx := strings.Index(text, " — "); idx >= 0 {
		return text[idx+3:]
	}
	for _, prefix := range []string{"Купон по ", "Погашение ", "Пут-оферта по "} {
		if strings.HasPrefix(text, prefix) {
			return text[len(prefix):]
		}
	}
	return text
}

var mergeableKinds = map[string]bool{"coupon": true, "maturity": true, "put_offer": true, "purchase": true}

func refreshMergedDescription(e *CashflowEvent) {
	if e.RelatedISIN == nil {
		return
	}
	name := bondNameFromDescription(e.Description)
	if e.Kind == "purchase" {
		e.Description = CashflowEventDescription("purchase", name, e.BondsCount, e.Lots, "")
		return
	}
	priceSuffix := ""
	if e.Kind == "put_offer" && strings.Contains(e.Description, " (") {
		tail := strings.SplitN(e.Description, " (", 2)[1]
		if strings.Contains(tail, "% номинала)") {
			priceSuffix = " (" + strings.Split(tail, ")")[0] + ")"
		}
	}
	e.Description = CashflowEventDescription(e.Kind, name, e.BondsCount, nil, priceSuffix)
}

type mergeKey struct {
	date time.Time
	kind string
	isin string
}

func MergeCashflowEvents(events []CashflowEvent) []CashflowEvent {
	sorted := sortedJournal(events)
	merged := make(map[mergeKey]*CashflowEvent)
	var order []mergeKey
	var passthrough []CashflowEvent
	for _, event := range sorted {
		if !mergeableKinds[event.Kind] || event.RelatedISIN == nil {
			passthrough = append(passthrough, event)
			continue
		}
		key := mergeKey{event.Date, event.Kind, *event.RelatedISIN}
		existing, ok := merged[key]
		if !ok {
			copy := event
			merged[key] = &copy
			order = append(order, key)
			continue
		}
		existing.AmountRub += event.AmountRub
		existing.IsProjected = existing.IsProjected || event.IsProjected
		if event.JournalSeq > 0 && (existing.JournalSeq == 0 || event.JournalSeq < existing.JournalSeq) {
			existing.JournalSeq = event.JournalSeq
		}
		if event.Lots != nil {
			v := derefInt(existing.Lots) + *event.Lots
			existing.Lots = &v
		}
		if event.BondsCount != nil {
			v := derefInt(existing.BondsCount) + *event.BondsCount
			existing.BondsCount = &v
		}
		refreshMergedDescription(existing)
	}
	var result []CashflowEvent
	for _, key := range order {
		result = append(result, *merged[key])
	}
	result = append(result, passthrough...)
	return sortedJournal(result)
}

func derefInt(p *int) int {
	if p == nil {
		return 0
	}
	return *p
}

func slotSortKey(slot ReinvestmentSlot) (time.Time, int, string) {
	reasonOrder := map[ReinvestmentTriggerReason]int{
		TriggerMaturity: 0, TriggerPutOffer: 1, TriggerCouponCash: 2,
	}
	src := ""
	if slot.SourcePositionISIN != nil {
		src = *slot.SourcePositionISIN
	} else if eff := slot.EffectiveISIN(); eff != nil {
		src = *eff
	}
	return slot.TriggerDate, reasonOrder[slot.TriggerReason], src
}

func copySlot(slot ReinvestmentSlot) ReinvestmentSlot {
	return ReinvestmentSlot{
		TriggerDate: slot.TriggerDate, TriggerReason: slot.TriggerReason,
		ExpectedCashRub: slot.ExpectedCashRub, SuggestedISIN: slot.SuggestedISIN,
		SuggestedName: slot.SuggestedName, ConfirmedISIN: slot.ConfirmedISIN,
		GapDays: slot.GapDays, SourcePositionISIN: slot.SourcePositionISIN,
	}
}

func accumulateSlot(existing, slot *ReinvestmentSlot) {
	existing.ExpectedCashRub += slot.ExpectedCashRub
	if existing.ConfirmedISIN == nil {
		existing.ConfirmedISIN = slot.ConfirmedISIN
	}
	if existing.SuggestedISIN == nil {
		existing.SuggestedISIN = slot.SuggestedISIN
	}
	if existing.SuggestedName == nil {
		existing.SuggestedName = slot.SuggestedName
	}
	priority := map[ReinvestmentTriggerReason]int{TriggerMaturity: 0, TriggerPutOffer: 1, TriggerCouponCash: 2}
	if priority[slot.TriggerReason] < priority[existing.TriggerReason] {
		existing.TriggerReason = slot.TriggerReason
		existing.TriggerDate = slot.TriggerDate
		if slot.SourcePositionISIN != nil {
			existing.SourcePositionISIN = slot.SourcePositionISIN
		}
	} else {
		if slot.TriggerDate.Before(existing.TriggerDate) {
			existing.TriggerDate = slot.TriggerDate
		}
		if existing.SourcePositionISIN == nil {
			existing.SourcePositionISIN = slot.SourcePositionISIN
		}
	}
}

func mergeSlotGroups(slots []ReinvestmentSlot, keyFn func(ReinvestmentSlot) string) []ReinvestmentSlot {
	sorted := append([]ReinvestmentSlot(nil), slots...)
	sort.Slice(sorted, func(i, j int) bool {
		di, oi, si := slotSortKey(sorted[i])
		dj, oj, sj := slotSortKey(sorted[j])
		if di.Equal(dj) {
			if oi == oj {
				return si < sj
			}
			return oi < oj
		}
		return di.Before(dj)
	})
	merged := make(map[string]*ReinvestmentSlot)
	var order []string
	var passthrough []ReinvestmentSlot
	for _, slot := range sorted {
		key := keyFn(slot)
		if key == "" {
			passthrough = append(passthrough, slot)
			continue
		}
		if existing, ok := merged[key]; ok {
			accumulateSlot(existing, &slot)
			continue
		}
		copy := copySlot(slot)
		merged[key] = &copy
		order = append(order, key)
	}
	var result []ReinvestmentSlot
	for _, key := range order {
		result = append(result, *merged[key])
	}
	result = append(result, passthrough...)
	sort.Slice(result, func(i, j int) bool {
		di, oi, si := slotSortKey(result[i])
		dj, oj, sj := slotSortKey(result[j])
		if di.Equal(dj) {
			if oi == oj {
				return si < sj
			}
			return oi < oj
		}
		return di.Before(dj)
	})
	return result
}

func slotMergeKey(slot ReinvestmentSlot) string {
	if slot.TriggerReason == TriggerCouponCash {
		return slot.TriggerDate.Format("2006-01-02") + "|" + string(slot.TriggerReason)
	}
	if slot.SourcePositionISIN != nil {
		return slot.TriggerDate.Format("2006-01-02") + "|" + string(slot.TriggerReason) + "|" + *slot.SourcePositionISIN
	}
	return ""
}

func slotCoalesceKey(slot ReinvestmentSlot) string {
	eff := slot.EffectiveISIN()
	if eff == nil {
		return ""
	}
	return slot.PurchaseDate().Format("2006-01-02") + "|" + *eff
}

func MergeReinvestmentSlots(slots []ReinvestmentSlot) []ReinvestmentSlot {
	bySource := mergeSlotGroups(slots, slotMergeKey)
	return mergeSlotGroups(bySource, slotCoalesceKey)
}
